pilot/pkg/model/context.go

```diff
@@ -119,6 +120,223 @@ var (
 	istioVersionRegexp = regexp.MustCompile(`^([1-9]+)\.([0-9]+)(\.([0-9]+))?`)
 )
 
+// StringList is a list that will be marshaled to a comma separate string in Json
+type StringList []string
+
+func (l StringList) MarshalJSON() ([]byte, error) {
+	if l == nil {
+		return nil, nil
+	}
+	return []byte(`"` + strings.Join(l, ",") + `"`), nil
+}
+
+func (l *StringList) UnmarshalJSON(data []byte) error {
+	if len(data) == 0 || string(data) == `""` {
+		*l = []string{}
+	} else {
+		*l = strings.Split(string(data[1:len(data)-1]), ",")
+	}
+	return nil
+}
+
+// PodPort describes a mapping of port name to port number. Generally, this is just the definition of
+// a port in Kubernetes, but without depending on Kubernetes api.
+type PodPort struct {
+	// If specified, this must be an IANA_SVC_NAME and unique within the pod. Each
+	// named port in a pod must have a unique name. Name for the port that can be
+	// referred to by services.
+	// +optional
+	Name string `json:"name,omitempty"`
+	// Number of port to expose on the pod's IP address.
+	// This must be a valid port number, 0 < x < 65536.
+	ContainerPort int `json:"containerPort"`
+	// Name of the protocol
+	Protocol string `json:"protocol"`
+}
+
+type PodPortList []PodPort
+
+func (l PodPortList) MarshalJSON() ([]byte, error) {
+	if l == nil {
+		return nil, nil
+	}
+	b, err := json.Marshal([]PodPort(l))
+	if err != nil {
+		return nil, err
+	}
+	b = bytes.ReplaceAll(b, []byte{'"'}, []byte{'\\', '"'})
+	out := append([]byte{'"'}, b...)
+	out = append(out, '"')
+	return out, nil
+}
+
+func (l *PodPortList) UnmarshalJSON(data []byte) error {
+	var pl []PodPort
+	pls, err := strconv.Unquote(string(data))
+	if err != nil {
+		return nil
+	}
+	if err := json.Unmarshal([]byte(pls), &pl); err != nil {
+		return err
+	}
+	*l = pl
+	return nil
+}
+
+// NodeMetadata defines the metadata associated with a proxy
+// Fields should not be assumed to exist on the proxy, especially newly added fields which will not exist
+// on older versions.
+type NodeMetadata struct {
```

Comment:
do we want a notice about not changing the `json` names and/or keeping them in sync with things like `EXCHANGE_KEYS` somehow?